cilium-health/launch/endpoint.go

```diff
@@ -0,0 +1,211 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package launch
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/common/addressing"
+	"github.com/cilium/cilium/common/plugins"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/ipam"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/vishvananda/netlink"
+)
+
+func logFromCommand(cmd *exec.Cmd, netns string) error {
+	scopedLog := log.WithField("netns", netns)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stdout)
+		for in.Scan() {
+			scopedLog.Debugf("%s", in.Text())
+		}
+	}()
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stderr)
+		for in.Scan() {
+			scopedLog.Infof("%s", in.Text())
+		}
+	}()
+
+	return nil
+}
+
+func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing) error {
+	routes := []plugins.Route{}
+	v4Routes, err := plugins.IPv4Routes(addressing)
+	if err == nil {
+		routes = append(routes, v4Routes...)
+	} else {
+		log.Debugf("Couldn't get IPv4 routes for health routing")
+	}
+	v6Routes, err := plugins.IPv6Routes(addressing)
+	if err != nil {
+		return fmt.Errorf("Failed to get IPv6 routes")
+	}
+	routes = append(routes, v6Routes...)
+
+	prog := "ip"
+	args := []string{"netns", "exec", netns, "bash", "-c"}
+	buf := bytes.Buffer{}
+	join := "&&"
+	for _, rt := range routes {
+		cmd := strings.Join(rt.ToIPCommand(dev), " ")
+		log.WithField("netns", netns).WithField("command", cmd).Info("Adding route")
+		buf.WriteString(cmd)
+		buf.WriteString(join)
+	}
+	buf.Truncate(buf.Len() - len(join))
+	args = append(args, buf.String())
+
+	log.Infof("Running \"%s %+v\"", prog, args)
+	out, err := exec.Command(prog, args...).CombinedOutput()
+	if err == nil {
+		log.Warn(out)
+	}
+
+	return err
+}
+
+// LaunchAsEndpoint launches the cilium-health agent in a nested network
+// namespace and attaches it to Cilium the same way as any other endpoint,
+// but with special reserved labels.
+func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressing, opts *option.BoolOptions) context.CancelFunc {
+	ip4, ip6, err := ipam.AllocateNext("")
+	if err != nil {
+		log.WithError(err).Fatal("Error while allocating cilium-health IP")
+	}
+
+	// XXX: Make the above IPs known to pkg/node
+
+	// XXX: Release IPs later on?
```

Comment:
Yes please

Reply:
These IPs are now managed in `pkg/node`.

They're allocated once on daemon startup, and never released, just like the other IPs reserved for daemon use like the loopback IP.